File: src/ca.erl
-module(ca).
-compile(export_all).
-behaviour(application).
-behaviour(supervisor).
-export([start/2, stop/1, init/1]).

start(_StartType, _StartArgs) ->
   R = cowboy_router:compile([{'_', [{"/", ca_enroll, []}]}]),
   {ok, _} = cowboy:start_clear(http,[{port,8046}],#{env => #{dispatch => R}}),
   supervisor:start_link({local, ?MODULE}, ?MODULE, []).
stop(_State) -> ok.
init([]) -> {ok, { {one_for_one, 5, 10}, []} }.

File: lib/ca.ex
defmodule CA do
  use Application
  use Supervisor

  require Record
  Enum.each(Record.extract_all(from_lib: "ca/include/PKIXCMP-2009.hrl"),
            fn {name, definition} -> Record.defrecord(name, definition) end)
  def init([]), do: {:ok, { {:one_for_one, 5, 10}, []} }

  def start(_type, _args) do
      :logger.add_handlers(:ldap)
      CA.CMP.start
      :supervisor.start_link({:local, __MODULE__}, __MODULE__, [])
  end
end
